src/mapreduce.jl

```diff
@@ -63,7 +63,12 @@ end
         sync_threads()
         index = 2 * d * (thread-1) + 1
         if index <= threads
-            @inbounds shared[index] = op(shared[index], shared[index+d])
+            other_val = if index + d <= threads
+                shared[index+d]
+            else
+                neutral
+            end
+            @inbounds shared[index] = op(shared[index], other_val)
```

Comment:
This uses `shared[index]` to accumulate values. As shared memory bandwidth is limited, it would be faster to use a register for this, and to only write the last accumulated value into shared memory, i.e. to write only just before the thread becomes inactive.

It would also make sense to eliminate neutral operations, i.e. to pull the `shared[index] = op(shared[index], other_val)` into the if statement.

Reply:
Yeah, the implementation is not very optimized. The neutral value is a pretty messy thing in its entirety, but harder to eliminate from the shuffle-based implementation.

Reply:
One reason I did it this way, IIRC, is that `op(val, neutral)` might change the element type; the output container element type (which we infer from `op`) might be different than the input.